File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <set>
using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
    vector<int> left(topping.size(),0);
    vector<int> right(topping.size(), 0);
    set<int> leftsize;
    set<int> rightsize;
    
    for (int i = 0; i < topping.size(); i++)
    {
        leftsize.insert(topping[i]);
        left[i] = leftsize.size();
    }
    for (int i = topping.size() - 1; i >= 0; i--)
    {
        rightsize.insert(topping[i]);
        right[i] = rightsize.size();
    }


    for (int i = 0; i < topping.size()-1; i++)
    {
        if (left[i] == right[i+1]) answer++;
    }
    return answer;
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cs
using System;
using System.Linq;
using System.Collections.Generic;
public class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        HashSet<int> left = new HashSet<int>();
        HashSet<int> right = new HashSet<int>();
        int[] leftsize = new int[topping.Length];
        int[] rightsize = new int[topping.Length];

        for(int i = 0; i < topping.Length; i++)
        {
            left.Add(topping[i]);
            leftsize[i] = left.Count;
        }
        for (int i = topping.Length -1; i >= 0; i--)
        {
            right.Add(topping[i]);
            rightsize[i] = right.Count;
        }

        for(int i = 0; i<topping.Length-1;i++)
        {
            if (leftsize[i] == rightsize[i+1])
            {
                answer++;
            }
        }
        return answer;
    }
}